lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp

```diff
@@ -0,0 +1,378 @@
+
+    if (clazz == options->augmentedGroundType) {
+      auto defName = anno.getAs<StringAttr>("defName").getValue();
+      auto name = anno.getAs<StringAttr>("name").getValue();
+
+      // TODO: This is ignoring situations where the leaves of and interface are
+      // not ground types.  This enforces the requirement that this runs after
+      // LowerTypes.  However, this could eventually be relaxed.
+      if (!type.isGround()) {
+        // TODO: This diagnostic doesn't print as report_fatal_error causes a
+        // very quick exit. Fix this.
+        auto diag = op->emitOpError()
+                    << "cannot be added to interface '" << defName
+                    << "', component '" << name
+                    << "' because it is not a ground type. (Got type '" << type
+                    << "'.) This will be dropped from the interface. (Did you "
+                       "forget to run LowerTypes?)";
+        diag.attachNote()
+            << "The annotation indicating that this should be added was: '"
+            << a;
+        llvm::report_fatal_error(
+            "Grand Central needs to run after Lower Types. A "
+            "GroundTypeAnnotation was found on a non-ground type.");
+      }
+
+      auto component = &(*interfaceMap)[{defName, name}];
+      component->found |= true;
+
+      switch (component->tpe) {
+      case Component::Type::Bundle:
+        // Do nothing.
+        break;
+      case Component::Type::Vector:
+        component->width = type.getBitWidthOrSentinel();
+        component->depth++;
+        break;
+      case Component::Type::Ground:
+        component->width = type.getBitWidthOrSentinel();
+        break;
+      case Component::Type::Unknown:
+        llvm_unreachable("Shouldn't be here");
+      }
+
+      continue;
+    }
+
+    unprocessedAnnos.push_back(a);
+  }
+
+  if (unprocessedAnnos.empty())
+    return {};
+  else
+    return ArrayAttr::get(context, unprocessedAnnos);
+};
+
+/// Combined logic to handle both Wires, Registers, and Nodes because these all
+/// use the same approach.
+void GrandCentralVisitor::handleRef(mlir::Operation *op) {
+
+  auto remainingAnnotations = handleRefLike(
+      op, getAnnotations(op), op->getResult(0).getType().cast<FIRRTLType>());
+
+  if (!remainingAnnotations || remainingAnnotations.empty())
+    op->removeAttr("annotations");
+  else
+    op->setAttr("annotations", remainingAnnotations);
+};
+
+/// Remove Grand Central Annotations from ports of modules or external modules.
+/// Return argument attributes with annotations removed.
+ArrayAttr GrandCentralVisitor::handlePorts(Operation *op) {
+
+  SmallVector<Attribute> newArgAttrs;
+  auto ports = getModulePortInfo(op);
+  for (size_t i = 0, e = ports.size(); i != e; ++i) {
+    auto port = ports[i];
+    auto argAttr = mlir::function_like_impl::getArgAttrs(op, i);
+    if (!port.annotations) {
+      newArgAttrs.push_back(DictionaryAttr::get(op->getContext(), argAttr));
+      continue;
```

Comment:
FYI I asked @youngar to ask some helpers to `AnnotationSet` to make working with port annotations easier/more self contained.

Reply:
This would be great! I was thinking that `AnnotationSet` is a little mismatched for ports. :100: